### PR TITLE
Implement custom sapcontrol webservices 

### LINKED ISSUE #79: Implement custom sapcontrol webservices
Currently sapcontrol code is imported from https://github.com/SUSE/sap_host_exporter/tree/master/lib/sapcontrol

The exporter only uses some specific `sapcontrol` functions, just for reading purposes. For trento the purpose might be different. 

In order to decouple trento from the prometheus exporter, the webservices code should be reimplemented in trento. The best location could be `trento/internal/sapsystems/sapcontrol`.

In this new implementation we can use `mapstructure` annotations to make the KV usage more standars (only using lower case for example), and remove the `viper` dependency, to number some of the improvements.

Finally, we could use `mockery` to mock the sapcontrol mocks and remove `gomock` dependecy.

### PR DESCRIPTION
**Pull request by fabriziosestito**
This PR decouples trento from [sap_host_exporter](https://github.com/SUSE/sap_host_exporter/), introducing `trento/internal/sapsystems/sapcontrol` containing a stripped-down version of the `gowsdl` generated soap webservice.
Here the `sap_host_exporter` original code for reference: https://github.com/SUSE/sap_host_exporter/tree/master/lib/sapcontrol

Webservice and soap client initialization has been moved away from `internal/sapsystem` to the webservice constructor, since we are just using a unix socket here, removing the need to use a `viper` configuration object.
Adds `mapstructure` tags to webservice result structs.

It also refactors tests by removing `sap_host_exporter` dependency `gomock`, using the `mockery` generator to be consistent with the rest of the project.

Resolves #79

Some questions:
should we add and link the full generated `gowsdl` code as reference to the documentation (see for example here: https://github.com/SUSE/sap_host_exporter/blob/master/doc/development.md#sapcontrol-web-service)?

Are the generated json/xml tags  in the webservice's structs necessary?

### CODE REVIEW

File: internal/sapsystem/sapcontrol/webservice.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+type InstanceProperty struct {
+	Property     string `xml:"property,omitempty" json:"property,omitempty" mapstructure:"property,omitempty"`
+	Propertytype string `xml:"propertytype,omitempty" json:"propertytype,omitempty" mapstructure:"propertytype,omitempty"`
+	Value        string `xml:"value,omitempty" json:"value,omitempty" mapstructure:"value,omitempty"`
+}
+
+type SAPInstance struct {
+	Hostname      string     `xml:"hostname,omitempty" json:"hostname,omitempty" mapstructure:"hostname,omitempty"`
+	InstanceNr    int32      `xml:"instanceNr,omitempty" json:"instanceNr,omitempty" mapstructure:"instanceNr,omitempty"`
```
Comment (arbulu89) [inline internal/sapsystem/sapcontrol/webservice.go:81]: In `mapstructure` annotations, to keep consistency, I would rename the values to use only lowercase.
Example: `instanceNr` -> `Instancenr` and `httpPort` -> `httpport`

File: internal/sapsystem/sapcontrol/webservice.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+type STATECOLOR string
+type STATECOLOR_CODE int
+
+const (
+	STATECOLOR_GRAY        STATECOLOR      = "SAPControl-GRAY"
+	STATECOLOR_GREEN       STATECOLOR      = "SAPControl-GREEN"
+	STATECOLOR_YELLOW      STATECOLOR      = "SAPControl-YELLOW"
+	STATECOLOR_RED         STATECOLOR      = "SAPControl-RED"
+	STATECOLOR_CODE_GRAY   STATECOLOR_CODE = 1
```
Comment (arbulu89) [inline internal/sapsystem/sapcontrol/webservice.go:30]: I think we don't use these `int` values by now. Maybe we should still keep them, but we could add a comment on top of them saying that they were just carried over form the `sap_host_exporter` project

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Hi @fabriziosestito . Fantastic work!
Just added some few details.

About your questions. I think the `xml` annotations are needed to the marshalling, I'm not sure about the `json` ones. @stefanotorresi will know for sure.
If we can remove them, let's remove them.

File: internal/sapsystem/sapsystem_kvstore_test.go [hunk 3]
```diff
@@ -5,19 +5,18 @@ import (
...
-	"github.com/SUSE/sap_host_exporter/lib/sapcontrol"
-	"github.com/SUSE/sap_host_exporter/test/mock_sapcontrol"
-	"github.com/golang/mock/gomock"
 	consulApi "github.com/hashicorp/consul/api"
 	"github.com/stretchr/testify/assert"
 	"github.com/trento-project/trento/internal/consul"
-	"github.com/trento-project/trento/internal/consul/mocks"
+	consulMocks "github.com/trento-project/trento/internal/consul/mocks"
+	"github.com/trento-project/trento/internal/sapsystem/sapcontrol"
+	sapcontroMocks "github.com/trento-project/trento/internal/sapsystem/sapcontrol/mocks"
```
Comment (rtorrero) [inline internal/sapsystem/sapsystem_kvstore_test.go:13]: I guess you meant `sapcontrolMocks` here?

**Comment by rtorrero** [comment]
Great work @fabriziosestito ! Thanks for the PR, I didn't find anything worth commenting other than the small typo on `internal/sapsystem/sapsystem_kvstore_test.go ` and what was already mentioned by @arbulu89 .

**Review by arbulu89 — DISMISSED** [review]
@fabriziosestito Green light from my side. If you tested the changes in a real trento env, and it still works the same as before, we can merge

**Comment by fabriziosestito** [comment]
> @fabriziosestito Green light from my side. If you tested the changes in a real trento env, and it still works the same as before, we can merge

I was not sure how I should confidence-test against a real trento env, so I basically exported the contents of the consul kv store from the master/this branch and compared them.

[master.json](https://github.com/trento-project/trento/files/6643704/master.txt)
[custom_sapcontrol_webservices.json](https://github.com/trento-project/trento/files/6643693/custom_sapcontrol_webservices.txt)

Comparing the two shows that only the key names and time-related values are changed.

Here the `diff -u` output for reference:
[diff.output.txt](https://github.com/trento-project/trento/files/6643742/diff.output.txt)


You might need to re-approve the PR since I squashed the fixup commits.

**Review by arbulu89 — APPROVED** [review]

**Comment by rtorrero** [comment]
LGTM

**Comment by stefanotorresi** [comment]
great stuff!

**Pull request merged**